fri/src/verifier/mod.rs

```diff
@@ -368,3 +497,83 @@ fn get_query_values<E: FieldElement, const N: usize>(
 
     result
 }
+
+fn iterate_through_query<B, E, H, const N: usize>(
+    layer_commitments: &Vec<H::Digest>,
+    folding_roots: &Vec<B>,
+    layer_alphas: &Vec<E>,
+    query: &Vec<(Vec<<H>::Digest>, [E; N])>,
+    position: usize,
+    num_partitions: usize,
+    folding_factor: usize,
+    number_of_layers: usize,
+    domain_size: usize,
+    evaluation: &E,
+    domain_generator: B,
+    max_degree_plus_1: usize,
+    domain_offset: B,
+) -> Result<(usize, E, usize, usize), VerifierError>
+where
+    B: StarkField,
+    E: FieldElement<BaseField = B>,
+    H: ElementHasher<BaseField = B>,
+{
+    let mut cur_pos = position;
+    let mut evaluation = *evaluation;
+    let mut domain_size = domain_size;
+    let mut domain_generator = domain_generator;
+    let mut max_degree_plus_1 = max_degree_plus_1;
+    let mut num_hash = 0usize;
+
+    for depth in 0..number_of_layers {
+        //println!("Depth is {:?}",depth);
+        let target_domain_size = domain_size / folding_factor;
+        let (query_proof, query_values) = &query[depth];
+
+        let folded_pos = cur_pos % target_domain_size;
+        let position_index =
+            map_position_to_index(&folded_pos, domain_size, folding_factor, num_partitions);
+
+        MerkleTree::<H>::verify(layer_commitments[depth], position_index, &query_proof)
+            .map_err(|_| VerifierError::LayerCommitmentMismatch)?;
+        num_hash += query_proof.len() - 1;
+
+        let query_value = query_values[cur_pos / target_domain_size];
```

Comment:
I think this is the part where we'd need to use the `AdviceSet` to read the query value from it.

Reply:
I have now implemented a draft of the ideas we discussed. Let me know what you think.
I have merged the unhashing of the values to a node into the `get_node` method. I can extracted into a separate method once we agree on the design.
Next step is to implement concrete versions of the verification methods for different folding factors and field extension degrees. I propose that we start with the one we discussed, namely `N == 2 & d == 2` .